File: JS/56.merge-intervals.js
/*
 * @lc app=leetcode id=56 lang=javascript
 *
 * [56] Merge Intervals
 */

// @lc code=start
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
  let res = [];
  if (!intervals || intervals.length === 0) return res;
  intervals.sort((a, b) => a[0] - b[0]);
  // console.log(intervals);
  let sol = intervals[0];
  res.push(sol);
  for (let interval of intervals) {
    if (sol[1] < interval[0]) {
      sol = interval;
      res.push(sol);
    } else {
      // merge
      sol[1] = Math.max(sol[1], interval[1]); // also update the result
    }
  }
  return res;
};
// @lc code=end

File: Rust/56.merge-intervals.rs
/*
 * @lc app=leetcode id=56 lang=rust
 *
 * [56] Merge Intervals
 */

// @lc code=start
impl Solution {
    pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        use std::cmp;
        let mut result = Vec::new();
        if intervals.len() == 0 {
            return result;
        }
        let mut new_intervals = intervals;
        new_intervals.sort_by(|a, b| a[0].cmp(&b[0]));
        let mut sol = new_intervals[0].clone();
        for interval in new_intervals {
            if(sol[1] < interval[0]) {
                result.push(sol.clone());
                sol = interval;
                // result.push(sol.clone()); // Some algorithms on leetcode have the push here because the value in result can be modified by reference. 
            } else { // merge
                sol[1] = cmp::max(sol[1],interval[1]);
            }
        }
        result.push(sol.clone());
        result
    }
}
// @lc code=end

